xen/arch/x86/tpm.c

```diff
@@ -30,11 +30,25 @@ asm (
 #include "boot/defs.h"
 #include "include/asm/intel_txt.h"
 #include "include/asm/slaunch.h"
+#include "include/asm/x86-vendors.h"
 #ifdef __va
 #error "__va defined in non-paged mode!"
 #endif
 #define __va(x)     _p(x)
 
+/*
+ * Generic CPUID function
+ * clear %ecx since some cpus (Cyrix MII) do not set or clear %ecx
+ * resulting in stale register contents being returned.
+ */
+#define cpuid(_op,_eax,_ebx,_ecx,_edx)          \
```

Comment:
I would put this in `#ifndef cpuid`, in case header files are reorganized, and move it to `slaunch.h` - it doesn't really have anything to do with TPM.

Also, I wonder if it would be better to do `is_intel_cpu()` analogous to AMD version in `txt_early()` instead of assembly. It's all the same to me, but some people would prefer to review C than assembly, which may make upstreaming slightly faster.

Reply:
e0d4647bc42c812f8bbb12152e39414c9e48dd0e. Moved to C renaming `txt_early` to `slaunch_early`.